core/src/main/java/org/apache/iceberg/util/SnapshotUtil.java

```diff
@@ -307,7 +307,10 @@ public static Schema schemaFor(Table table, long snapshotId) {
       return schema;
     }
 
-    // TODO: recover the schema by reading previous metadata files
+    snapshot = table.snapshot(snapshot.parentId());
+    if (snapshot != null) {
+      return table.schemas().get(snapshot.schemaId());
```

Comment:
Is it possible that this can throw?

It would be great to add a test case that covers this situation.